### ISSUE TITLE
reportUnnecessaryCast not working with optional types

### ISSUE THREAD

**Issue by mgmira** [microsoft/pyright#5333]
**Describe the bug**
reportUnnecessaryCast seems to be broken when using optional types

**To Reproduce**
```python
# pyright: reportUnnecessaryCast=true


from __future__ import annotations

from typing import cast


def f(x: int | None):
    y = cast(int | None, x)  # unnecessary cast not reported
    return y


def get(x: int):
    y = cast(int, x)  # unnecessary cast reported
    return y
```

**Expected behavior**
Error should be reported in both cases above

**VS Code extension or command-line**
pyright 1.1.314

**Comment by erictraut** [comment]
Thanks for the bug report. My initial implementation of `reportUnnecessaryCast` worked only with simple class instance types, not other types like unions, instantiable classes, Any, TypeVars, callables, etc. I've extended it to work with other types as well. This will be included in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.315, which I just published. It will also be included in this week's insiders release of pylance.
